components/places/src/storage.rs

```diff
@@ -263,21 +270,311 @@ fn add_visit(db: &impl ConnExt,
+            ids_to_update.push(page.row_id);
+            db.execute_named_cached(insert_meta_sql,
+                                    &[(":row_id", &page.row_id),
+                                      (":change_delta", &page.sync_change_counter)])?;
+
+            result.insert(page.guid.clone(), OutgoingInfo::Record(HistoryRecord {
+                id: page.guid,
+                title: page.title,
+                hist_uri: page.url.to_string(),
+                sortindex: page.frecency,
+                ttl: HISTORY_TTL,
+                visits,
+            }));
+        }
+
+        // We need to update the sync status of these items now rather than after
+        // the upload, because if we are interrupted between upload and writing
+        // we could end up with local items with state New even though we
+        // uploaded them.
+        sql_support::each_chunk(&ids_to_update, |chunk, _| -> Result<()> {
+            db.conn().execute(&format!("UPDATE moz_places SET sync_status={status}
+                                 WHERE id IN ({vars})",
+                                vars = sql_support::repeat_sql_vars(chunk.len()),
+                                status = SyncStatus::Normal as u8),
+                              chunk)?;
+            Ok(())
+        })?;
+
+        Ok(result)
+    }
+
+    pub fn finish_outgoing(db: &Connection) -> Result<()> {
+        // So all items *other* than those above must be set to "not dirty"
+        // (ie, status=SyncStatus::Normal, change_counter=0). Otherwise every
+        // subsequent sync will continue to add more and more local visits
+        // until every visit we have is uploaded. And we only want to do it
+        // at the end of the sync because if we are interrupted, we'll end up
+        // thinking we have nothing to upload.
+        // BUT - this is potentially alot of rows! Because we want "NOT IN (...)"
+        // we can't do chunking and building a literal string with the ids seems
+        // wrong and likely to hit max sql length limits.
+        // So we use a temp table.
+        trace!("Updating all synced rows");
+        // XXX - is there a better way to express this SQL? Multi-selects
+        // doesn't seem ideal...
+        db.conn().execute_cached("
+            UPDATE moz_places
+            SET sync_change_counter = sync_change_counter -
+                (SELECT change_delta FROM temp_sync_updated_meta m WHERE moz_places.id = m.id)
+            WHERE id IN (SELECT id FROM temp_sync_updated_meta)
+            ", &[])?;
+
+        trace!("Updating all non-synced rows");
+        db.execute_all(&[&format!("UPDATE moz_places
+                                   SET sync_change_counter = 0, sync_status = {}
+                                   WHERE id NOT IN (SELECT id from temp_sync_updated_meta)",
+                                   (SyncStatus::Normal as u8)),
+                         "DROP TABLE temp_sync_updated_meta"])?;
```

Comment:
...And `DELETE FROM temp_sync_updated_meta` here?

Reply:
👍